### PR TITLE
EVA-1287 Pilot for import into accessioning service - corner cases (use genbank)

### PR DESCRIPTION
**Pull request by jmmut**
Use the genbank contig accessions if possible.

Before, the chromosome coordinates where used if available. If not, we used the contig ones. 

This PR now ensures the contig ones are used (and replaced by genbank ones), except in the case that they are not found in the assembly report and the chromosome ones are present. In this is the case, use the chromosome coordinates to get the genbank synonym.

### CODE REVIEW

File: eva-accession-import/src/test/java/uk/ac/ebi/eva/accession/dbsnp/processors/ContigReplacerProcessorTest.java [hunk 1]
```diff
@@ -38,61 +40,132 @@
...
     private ContigReplacerProcessor processor;
 
+    private ContigReplacerProcessor genbankProcessor;
+
+    @Rule
+    public ExpectedException thrown = ExpectedException.none();
+
     @Before
     public void setUp() throws Exception {
         String fileString = ContigMappingTest.class.getResource("/input-files/assembly-report/AssemblyReport.txt").toString();
```
Comment (cyenyxe) [inline eva-accession-import/src/test/java/uk/ac/ebi/eva/accession/dbsnp/processors/ContigReplacerProcessorTest.java]: Could some of the entries that are not used by the tests be deleted from the report? The different corner cases are interleaved and it makes hard to find them.

I think keeping some of the unused ones is useful to represent a realistic report, so that is why I'm in favor of not dropping them all.

File: eva-accession-import/src/test/java/uk/ac/ebi/eva/accession/dbsnp/processors/ContigReplacerProcessorTest.java [hunk 2]
```diff
@@ -38,61 +40,132 @@
...
         ContigMapping contigMapping = new ContigMapping(fileString);
```
Comment (cyenyxe) [inline eva-accession-import/src/test/java/uk/ac/ebi/eva/accession/dbsnp/processors/ContigReplacerProcessorTest.java]: I just noticed we don't have tests to check that at least the number of records in a `ContigMapping` is right after loading the assembly report.

File: eva-accession-import/src/test/java/uk/ac/ebi/eva/accession/dbsnp/processors/ContigReplacerProcessorTest.java [hunk 3]
```diff
@@ -38,61 +40,132 @@
...
+        long start2 = 200;
         SubSnpNoHgvs input = new SubSnpNoHgvs(SS_ID, RS_ID, REFERENCE_ALLELE, ALTERNATE_ALLELE, ASSEMBLY, "", "",
-                                              UCSC_1, START, UCSC_1, START,
-                                              DbsnpVariantType.SNV, Orientation.FORWARD, Orientation.FORWARD,
-                                              Orientation.FORWARD, true, true, true, true, null, null, TAXONOMY);
-        assertEquals(SEQNAME_1, processor.process(input).getContigName());
+                                              CHROMOSOME, start1, MISSING_CONTIG, start2, DbsnpVariantType.SNV,
+                                              Orientation.FORWARD, Orientation.FORWARD, Orientation.FORWARD, true, true,
+                                              true, true, null, null, TAXONOMY);
+        assertEquals(start1, processor.process(input).getContigStart());
```
Comment (cyenyxe) [inline eva-accession-import/src/test/java/uk/ac/ebi/eva/accession/dbsnp/processors/ContigReplacerProcessorTest.java]: Can we check that the contig name is also the appropriate one, for the sake of completeness?

File: eva-accession-import/src/main/java/uk/ac/ebi/eva/accession/dbsnp/processors/ContigReplacerProcessor.java [hunk 4]
```diff
@@ -25,31 +25,69 @@
...
 
-        if (contigSynonyms.isIdenticalGenBankAndRefSeq()) {
-            if (subSnpNoHgvs.getChromosome() != null) {
-                subSnpNoHgvs.setChromosome(contigSynonyms.getSequenceName());
+        if (isPresentInAssemblyReport(contigSynonyms)) {
+            replaceFromContigToGenbank(subSnpNoHgvs, contigSynonyms);
+        } else {
+            if (isPresentInAssemblyReport(chromosomeSynonyms)) {
+                replaceFromChromosomeToGenbank(subSnpNoHgvs, chromosomeSynonyms);
+            } else {
```
[the remark this answers is not in the document]
Comment (jmmut) [reply eva-accession-import/src/main/java/uk/ac/ebi/eva/accession/dbsnp/processors/ContigReplacerProcessor.java]: I was experimenting improving the readability. having variables forces to put the implementation of "present in assembly report" (checking for nulls) inside the method, which clutters it a bit.

File: eva-accession-import/src/test/java/uk/ac/ebi/eva/accession/dbsnp/processors/AssemblyCheckerProcessorTest.java [hunk 5]
```diff
@@ -111,13 +111,13 @@ private SubSnpNoHgvs newSubSnpNoHgvs(String chromosome, long chromosomeStart, St
 
     @Test
     public void validReferenceAlleleSeqNameFastaSeqName() throws Exception {
-        SubSnpNoHgvs input = newSubSnpNoHgvs(SEQNAME_1, START, null, 0, REFERENCE_ALLELE, DbsnpVariantType.DIV);
+        SubSnpNoHgvs input = newSubSnpNoHgvs(SEQNAME_1, START, SEQNAME_1, START, REFERENCE_ALLELE, DbsnpVariantType.DIV);
```
[the remark this answers is not in the document]
Comment (jmmut) [reply eva-accession-import/src/test/java/uk/ac/ebi/eva/accession/dbsnp/processors/AssemblyCheckerProcessorTest.java]: yes

File: eva-accession-import/src/main/java/uk/ac/ebi/eva/accession/dbsnp/processors/ContigReplacerProcessor.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (pabarcgar) [reply eva-accession-import/src/main/java/uk/ac/ebi/eva/accession/dbsnp/processors/ContigReplacerProcessor.java]: +1

File: eva-accession-import/src/test/java/uk/ac/ebi/eva/accession/dbsnp/processors/ContigReplacerProcessorTest.java [hunk 6]
```diff
@@ -38,61 +40,132 @@
...
-                                              DbsnpVariantType.SNV, Orientation.FORWARD, Orientation.FORWARD,
-                                              Orientation.FORWARD, true, true, true, true, null, null, TAXONOMY);
-        assertEquals(SEQNAME_1, processor.process(input).getContigName());
+    public void shouldThrowIfContigAndChromosomeAreMissingInAssemblyReport() throws Exception {
+        SubSnpNoHgvs input = newMockSubSnpNoHgvs(MISSING_CHROMOSOME, MISSING_CONTIG);
+        thrown.expect(IllegalStateException.class);
+        processor.process(input);
+    }
+
+    private SubSnpNoHgvs newMockSubSnpNoHgvs(String missingChromosome, String missingContig) {
```
[the remark this answers is not in the document]
Comment (pabarcgar) [reply eva-accession-import/src/test/java/uk/ac/ebi/eva/accession/dbsnp/processors/ContigReplacerProcessorTest.java]: +1
